### DISCUSSION TITLE
[C128] Why are these the Zero Page address definitions and can they be changed?

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1738] [Q&A] [answered]
Hi community, 

I am was able to pick up my mini project again and make nice progress. 
(The one in which I am aiming to be able to use regular C-function calls
to run any routine regardless in which c128 memory bank it is located)

Currently I run into some optimization issues for which I hope somebody might share some light. 

The multi-memory bank access routine is based on assembler and now I am using it within the CC65. 
So the challenge is that I run into the per-assigned definitions for the Zero page locations in CC65.
I am wondering if there is someone who knows why certain locations have been chosen.
Furthermore, what the impact is if they would be changed and where to change them. 

So when I look at the documentation (Mapping C128) and the CC65 compiler generated 'mapping file'
next to the C128.inc I think the mapping list looks like this;  
(Please correct me if I am wrong)

=== Mapfile.txt and Mapping C128================================================
Addr | CC65 | Mapping C128
-----|------|-------------
000002 | sp | BANK 	        Last Bank#Nr before BRK      			
000003 | [not assigned] | PC	        PC / SYS Address for JMPFAR & JSRFAR		
000004 | sreg | PC	        PC / SYS Address for JMPFAR & JSRFAR
000005 | [not assigned] | S-REG	         status Register for JMPFAR & JSRFAR
000006 | regsave | A-REG	         A-Register for JMPFAR & JSRFAR
000007 | [not assigned]	|   X-REG	         X-Register for JMPFAR & JSRFAR
000008 | [not assigned]	|   Y-REG	         Y-Register for JMPFAR & JSRFAR
000009 | [not assigned]	|   STKPTR	          sp (storage) 8502 system stack

00000A	ptr1
00000B
00000C	ptr2
00000D       
00000E  ptr3
00000F       
000010  ptr4
000011       

000012  tmp1       
000013  tmp2     
000014  tmp3     
000015  tmp4     

000016	regbank    			

0000FA [not assigned]
0000FB [not assigned]
0000FC [not assigned]
0000FD [not assigned]
0000FE [not assigned]

=== additional from C128.inc =================================================
TXTPTR             := $3D          ; Pointer into BASIC source code
TIME                := $A0          ; 60HZ clock
FNAM_LEN       := $B7          ; Length of filename
SECADR            := $B9          ; Secondary address
DEVNUM          := $BA          ; Device number
FNAM               := $BB          ; Address of filename
FNAM_BANK     := $C7          ; Bank for filename
KEY_COUNT      := $D0          ; Number of keys in input buffer
FKEY_COUNT    := $D1          ; Characters for function key
MODE               := $D7          ; 40/80 column mode flag
CURS_X             := $EC          ; Cursor column
CURS_Y             := $EB          ; Cursor row
SCREEN_PTR      := $E0          ; Pointer to current char in text screen
CRAM_PTR        := $E2          ; Pointer to current char in color RAM

===============================================================================
**My Questions are;** 
(Sorry that there seem to be a lot.......) 
**000002	sp**

- How is the sp in CC65 used, just like the system sp? 
- A counter growing down from from $FF? 
- Can it be relocated to a different address? 
- If, it can be relocated where is the address assigned/defined
- (where does it needs to be changed)

**000003  [not assigned]	PC	PC / SYS Address for JMPFAR & JSRFAR		
000004  sreg     	        PC	PC / SYS Address for JMPFAR & JSRFAR**

- What is the purpose of this value and can it be relocated?
- If, it can be relocated where is the address assigned/defined
- (where does it needs to be changed)

**000006  regsave** 

- Where is this regsave used for within the compiler? 

**000007	[not assigned]	X-REG	X-Register for JMPFAR & JSRFAR
000008	[not assigned]	Y-REG	Y-Register for JMPFAR & JSRFAR
000009	[not assigned]	STKPTR	sp (storage) 8502 system stack**

- As these don't seem to be assigned within the CC65 can I freely use them for my own purpose? 
- In which file are the tmp1-tmp4 and ptr1-ptr4 defined? 

**000016	regbank**   

- Where is this regsave used for within the compiler? 

**0000FA [not assigned]
0000FB [not assigned]
0000FC [not assigned]
0000FD [not assigned]
0000FE [not assigned]**

- Why are these know free locations not defined as available zero page 'tmp' variables?
- In which files should I define these if I want to 'formalize' them as CC65 'tmp' variables

Based on the answers I will know what I can change and where. Which will be most helpful. 
Thanks in advance!

**Comment by polluks** [comment]
"Luke, use the source!" the day after tomorrow :-)
At least the not assigned range may be reserved for assembler, for example an IRQ.

**Reply by RetrofanC** [reply]
(-8 Understood!

But based on the reply do I understand correct their is no historical information or documentation available why the choice made for the zero page addresses are what they are? As I have not been able to find that the standard CC65 user materials unfortunately. 
(I was hoping there was some administrator materials perhaps or something)

**Comment by mrdudz** [comment]
I think we are looking at an XY Problem here. For a start, i dont see how it is a challenge to use that assembler routine - just make another zeropage segment in the linker config and use that in the assembler code.

Then, cc65 has some support builtin to make banked calls (although not fully automatic), which is likely the preferred way to do something like this.

That said, generally most things you want to do boil down to making a custom linker config

**Reply by RetrofanC** [reply]
Thanks for the reply and could be XY indeed I am not sure, but to your point; 

"just make another zeropage segment"

Do you mean I make my own ZP cfg with my own definitions and replace that for the current one? 

That would be ok but than still could I assume what has not been named/defined/allocated  in the default ZP definition is not used within CC65? As I want to avoid defining a new one and get 'strange' behaviors by the fact CC65 is actually using zp address locations which are just not documented or defined in the default zp.  

As to your second point of CC65 Banking. I have understood that what CC65 defines as 'banking' is different as what is needed. 
CC65 banking means it is just 'swapping' memory sections. So there is only one 64k memory in which code is actually executed. 
However, what I need is that it is possible to actually 'execute' code into a different memory bank. This requirement and capability is rather specific to the C128 it seems and I want make it available through CC65 as well. 

Agreed on the custom linker config so I am looking to as much as details and information on how things are defined and working in CC65/C128. So I can make the required changes to make it all work but keep the CC65 as stock as possible. As the end goals would be to donate it to CC65 as default (or accepted add-on) capability. 

So any further additional information or suggestions are welcome!

**Reply by mrdudz** [reply]
> However, what I need is that it is possible to actually 'execute' code into a different memory bank. This requirement and capability is rather specific to the C128 it seems and I want make it available through CC65 as well.

Actually it does have support for this - you can mark functions in header files as banked functions, and then the compiler will produce code that calls them via a banking trampoline which resides somewhere in non banked memory. I am using this a lot actually (and infact writing a proper example program for this has been on my todo list for a while).

**Comment by mrdudz** [comment]
> But based on the reply do I understand correct their is no historical information or documentation available why the choice made for the zero page addresses are what they are?

I don't think so. I think what happened was that when Uz started the project he created a linker config for the C64 and just used the first N zeropage locations, starting with 2 because 0/1 is used for the processor port. Then further linker configs for other systems were probably copies of the C64 config and then adjusted to work.

The only real constraint for what zeropage locations are used is that they must be in one continues block of memory with the size of $18 bytes - see [here](https://github.com/cc65/cc65/blob/master/cfg/c128.cfg). There is no practical way to lift this limitation at the moment, ie you cant freely choose where each of the used locations are, you can only move the whole block.

That said, i'd still like to know what exactly you are trying to do here. Why are the currently choosen locations a problem?

**Reply by RetrofanC** [reply]
Thanks for the quick reply and valuable information! Our messages crossed just now sorry for that!

As to your question; 

In my code the last known memory bank is stored in $02 just like it would be on a stock C128. 
However, in CC65 this location seemed to be used for the CC65 'Stack Pointer'. 
So depending on what happens to the stack pointer value during a bank jump my application will crash. 

In a similar fashion currently $03 and $04 are used for the jump address to one of the other 3 memory banks. 
Given the CC65 seems to use $04 for the sreg this value could change during a jump or changes when the sreg is read by other pieces of the code. 
 
These are just two examples but that is the challenge. So if I could make the default CC65 zp config more inline with a stock C128.
I could avoid these challenges. 
Surely I can also change my code and make it dedicate for the CC65 to find available zp memory locations. But explaining what my code does is easier when it just uses the zp locations which are well documented for the stock C128 memory mapping. 

Does this help? 
Otherwise, more than happy to go into deeper as needed as I really appreciated you helping on this topic! 
As I realize it is asking for more (uncommon) details as normally needed.

**Reply by RetrofanC** [reply]
Btw good remark which I didn't look into close enough! 
There no zp defined of $ff but only $18.
So the other locations above that are not used as such. 

Any idea is that arbitrary?
So can I just extend that with a large value as long as it is one block or it should remain $18 ($1A-$02) fix length perhaps?

**Comment by mrdudz** [comment]
First, for your own zeropage variables i'd create an extra segment - which you can then place anywhere you want. That said, if you are a bit careful, you can reuse some of the existing zeropage variables (tmp1.. and ptr1..) in your own asm functions.

Now for implementing the banking trampolines, i wouldnt use zeropage locations to save/restore the bank - but instead push those values on the stack. Because that then allows any banked function to call any other banked function, and you can move functions from bank to another without having to worry about what is where.

**Comment by mrdudz** [comment]
Here is an example on how to implement banked functions:

```
; The "bank" parameter is passed via tmp4.
;
; The address of a wrapped function is passed in ptr4. The wrapper can call
; that function by using "jsr callptr4".
;
; The wrapper must preserve A/X until calling the wrapped function, and after
; it returns. It must also preserve Y when the wrapped function has variadic
; parameters (...)
;
; #pragma wrapped-call (push, overlaycall, bank)
; void somefunc1(void);
; #pragma wrapped-call (pop)

; void overlaycall(void)
_overlaycall:

    tay ; preserve A

    ; save current bank
    lda bankregister
    pha

    ; switch to new bank

    ; tmp4 contains the new bank
    lda tmp4
    sta bankregister

    tya ; restore A

    jsr callptr4        ; call wrapped function

    tay ; preserve A

    ; restore previous bank
    pla
    sta bankregister

    tya ; restore A
    rts
```

to call such banked function from assembler set up ptr4 and tmp4 before calling the above function manually:

```
    .export initwrappedcall
    .importzp ptr4
    .importzp tmp4

; uses to call banked functions from assembler like this:
; 
; lda #<_mybankedfunction
; ldx #>_mybankedfunction
; ldy #<.bank(_mybankedfunction)
; jsr initwrappedcall
; jsr _plugincall

initwrappedcall:
    sta ptr4        ; function lo
    stx ptr4 + 1    ; function hi
    sty tmp4        ; function bank
    rts
```

**Reply by RetrofanC** [reply]
Thanks, let me play with this a bit and see if I can make it work in this way!

**Reply by mrdudz** [reply]
> Thanks, let me play with this a bit and see if I can make it work in this way!

BTW, for this to work, you'll have to make a linker config that places the common code, the zeropage, the hardware stack, the software stack, and the heap in a non banked area. Most practical is probably to use eg the first 4k or 8k ram for this (and place the screen there as well)

**Reply by RetrofanC** [reply]
I made some nice progress over time and got my project almost working. 
So I can call a function in any RAM0/1/2/3 of the C128 and have them executing. 
I also have it working when the function has 1 variable (regardless if it is int or char) and the function can have a return value as well. All working fine. 

However, when I want to have more than one variables defined it doesn't work. 
Meaning I don't get any error messages but there are just no values in those variables when they arrive at the function. It is unclear if the "pushax" doesn't put the values in or if the "ldaxysp" can't get them out. 

The values are always 0xff when I read them in the called function. However, one strange thing is that if I define a int I get $ffff and when I define a char I get $00ff. So it looks something is happening....... 
 
As per your earlier suggestion in your message above I put all possible source files (.s) into the common ram ("BANKING") and defined a common ram of 4k. 

After trying a lot of things for several days I am a bit stuck.... 
Hereby the map and (linker generated) asm file which perhaps might provide the first clues of what I am doing wrong. 

[mapfile.txt](https://github.com/cc65/cc65/files/10006399/mapfile.txt)
[simpleswitch_asm.txt](https://github.com/cc65/cc65/files/10006511/simpleswitch_asm.txt)

I replied to this message as you were the one being very clear on what I had to put into the common ram. 
If you could provide any guidance or suggestions that would be great. 

If anybody else has an idea or suggestion please let me know as well.

I know I am almost there but what is the last piece of the puzzle......... 

(If there is any other document or source code needed just let me know)

**Reply by mrdudz** [reply]
It sounds to me like you'll have to relocate the software stack (which the runtime library of the compiler uses, eg to pass parameters to functions) into common RAM too. Have a look at https://github.com/cc65/cc65/blob/master/libsrc/c128/crt0.s#L59 - this is where it sets up the initial stack pointer (in that case, the default is the last address in available RAM). So what you need to do is:
- if you dont already do this for other reasons, make a copy of that crt0.s into your source and link that one
- find an area in common RAM where the stack can live. How much memory you will need there depends a lot on the type of program, for example i am using 512 bytes in one of my projects. 1k is probably more than enough for most things. Just find one block that you can reserve for this (*)
- Change the setup of sp in crt0.s so it points to the top (last address) of the memory block you have chosen for the stack. Since you are now using a custom crt0 anyway, you may just hardcode the address there for a test. To make it more clean change your linker config to produce a memory section for the stack, export the start and size of it, and then import those symbols in crt0 and calculate the sp address in a similar way it does now.

> If there is any other document or source code needed just let me know

Unfortunately no. In fact, when you get this working, i'd be really interested in seeing the whole setup - we might adapt some ideas for the regular c128 setup that is :) And perhaps we can copypaste some kind of documentation from this discussion :)

(*) edit: to get an idea of how much you need, fill the block with some known pattern ( $5a, $a5, etc ) before running the cc65 program (or just do it in crt0), and after running the program for a while in emulator check the stack area. leave a reasonable safety margin above what you can see it is using.

**Reply by RetrofanC** [reply]
Thanks soooo much you are my hero of the day! (-8 
It seems to do the trick indeed!!!

I will do some more testing having functions with more values than two and making more complex functions. 
But the first test with having the stack running down from $1000 which is the top of the Common RAM seems to work.

Yes, I am more than eager to give back to the community for sure. 
So when I got it tested and working fully I will clean up the code and make a nice description. 
Share it here first for all to test and take further guidance on how to best share further.

**Comment by Marrin** [comment]
I don't know if this is clear or not: the addresses of those zero page variables are not chosen per variable. There is no explicit definition that says `sp` goes to $02/$03. There is a memory range called `ZP` at $02 that is $1A bytes long and a segment `ZEROPAGE` that maps to that memory address. That `sp` maps to $02/$03 is just because `sp` is the first name that is assigned to a word in the ZEROPAGE segment. There are some pointers (`ptr1`…) and temporary word sized variables (`tmp1`…) there, a two byte variable for the 16 bit of 32 bit results that don't fit into `A` and `X`, space to save register values, and space to store arguments and local variables declared `register` variables in C.

The stack pointer `sp` is a *word* by the way and points to the software stack that is used to store arguments and local variables. It grows downwards. The hardware stack is used for return addresses; everything else goes into ”normal” memory.

**Reply by RetrofanC** [reply]
This is a good addition indeed as I was not fully realizing that as well, thanks. 

Surely SP should be a word but indeed I didn't realize that enough when writing the overview so good catch!  

Reading your explanation does it mean that indeed this ZEROPAGE segment could end up in any part of the memory and is not designated to the actual zero page memory range of the C128? 
(I means is somehow similar with the MMU zeropage relocation feature which allows to relocate the zeropage to 'any' part of the C128 memory?)

**Comment by mrdudz** [comment, marked as answer]
> does it mean that indeed this ZEROPAGE segment could end up in any part of the memory and is not designated to the actual zero page memory range of the C128?

well, you could change the linker config that way - but of course the code will no more work then, since certain address modes will only work with zeropage
